heaviside/activities.py

```diff
@@ -291,67 +372,137 @@ def heartbeat(self):
-        self.task_proc = task_proc
+                return activity['activityArn']
+        return None
 
-    def run(self):
-        """Called by Process.start() to process data
+    def create_worker_name(self):
+        """Generate a unique worker name"""
+        rand = ''.join(random.sample(CHARS, 6))
 
-        Note: Automatically creates the Activity ARN if it doesn't exist
+        if name is None:
+            name = self.name
 
-        """
-        self.create()
+        if name is not None:
+            rand = name + '-' + rand
 
-        get_worker = lambda: (self.name + '-' + ''.join(random.sample(CHARS,6)))
+        return rand
 
-        # Note: needed for unit test, so the loop will exit
-        self.running = True
-        while self.running:
+    def run(self, name=None):
+        """Start polling for Activity tasking
+
+        Args:
+            name (String|None) : Name or ARN of the Activity to monitor, if arn
+                                 is not already defined
+        """
+        if name is None:
+            if not hasattr(self, 'arn') or self.arn is None:
+                raise Exception("Could not locate Activity ARN to poll")
+
+            self.name = self.arn.split(':')[-1]
+        else:
+            if name.lower().startswith("arn:"):
+                self.arn = name
+                self.name = self.arn.split(':')[-1]
+            else:
+                self.arn = self.lookup_activity_arn(name)
+                self.name = name
+                self.create_activity(name)
+
+        # Storage for currently executing workers
+        self.workers = []
+
+        self.log.debug("Starting polling {} for tasking".format(self.arn))
+        # DP NOTE: needed for unit test, so the loop will exit
+        # DP TODO: make thread safe so it can be used to stop a running thread
+        self.polling = True
+        while self.polling:
             try:
-                worker = get_worker()
-                token, input_ = self.task(worker)
+                worker = self.create_worker_name()
+                token, input_ = self.poll_task(worker)
                 if token is not None:
                     try:
-                        proc = self.task_proc(worker, token, input_, **self.credentials)
-                        proc.start()
+                        results = self.handle_task(token, input_) # DP ???: What to do with worker name?
+
+                        # DP TODO: seperate error handling for concurrent processes from handle_task
+                        if results is not None:
+                            results.start()
+                            # DP TODO: Need a check to make sure the subprocess/thread starts
+                            #          There can be a problem if the subprocess/thread doesn't start
+                            #          or get to its own error handling logic
+                            # ???: Maybe store workers as token : subprocess, allowing us to always try to fail
```

Comment:
It will at least look similar. There are ways that a heartbeat can be done. Using the TaskProcesses should help prevent this. The `run` method for TaskProcess wraps the call to `handle_task` in a try/except and will try to fail the task if there was an error. Since this is the first code executed by the Process's `start` method it should catch any unexpected exceptions.